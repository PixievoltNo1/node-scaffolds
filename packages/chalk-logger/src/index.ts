import type { Mutable } from '@guanghechen/utility-types'
import { partOptionsFromArgs } from './commander'
import type { Level } from './level'
import type { ILoggerOptions } from './logger'
import { Logger } from './logger'

export * from './color'
export * from './commander'
export * from './level'
export * from './logger'

export class ChalkLogger extends Logger {
  /**
   * prefix of logger.name
   */
  protected basename: string | null = null
  /**
   * name passed into .setName()
   */
  protected divisionName: string | null = null

  constructor(options?: ILoggerOptions, args?: string[]) {
    const optionsFromArgs = partOptionsFromArgs(args || [])
    const resolvedOptions = {
      ...options,
      ...optionsFromArgs,
      flags: { ...options?.flags, ...optionsFromArgs.flags },
    }

    super(resolvedOptions)
    const basename = resolvedOptions?.name ?? ''
    this.setBaseName(basename)
  }

  /**
   * update logger's level
   * @param level
   */
  public setLevel(level: Level | null | undefined): void {
    if (level == null) return
    const self = this as Mutable<this>
    self.level = level
  }

  /**
   * update logger's name
   * @param name
   */
  public setName(name: string | null): void {
    const resolvedName: string = [this.basename, name]
      .filter((x): x is string => x != null && x.length > 0)
      .join(' ')
    const self = this as Mutable<this>
    self.name = resolvedName
    this.divisionName = name
  }

  /**
   * update basename of logger
   * @param basename
   */
  public setBaseName(basename: string | null): void {
    this.basename = basename
    this.setName(this.divisionName)
  }

  /**
   * update logger's mode
   * @param mode
   */
  public setMode(mode: 'normal' | 'loose'): void {
    const self = this as Mutable<this>
    self.mode = mode
  }
}

export default ChalkLogger
