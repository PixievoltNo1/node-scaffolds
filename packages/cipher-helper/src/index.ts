export * from './_base'
export * from './aes'
export * from './catalog'
export * from './types/catalog'
export * from './types/cipher-helper'
export * from './util/buffer'
export * from './util/key'
