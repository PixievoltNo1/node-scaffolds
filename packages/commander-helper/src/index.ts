export * from './command'
export * from './fs'
export * from './git'
export * from './merge'
export * from './option'
export * from './yarn'
