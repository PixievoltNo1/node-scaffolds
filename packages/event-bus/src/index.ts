export * from './SimpleEventBus'
export * from './types'
