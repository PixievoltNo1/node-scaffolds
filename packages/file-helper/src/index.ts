export * from './big-file'
export * from './util'
export * from './types'
