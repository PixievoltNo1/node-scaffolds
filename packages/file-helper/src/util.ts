import type { ChalkLogger } from '@guanghechen/chalk-logger'
import invariant from '@guanghechen/invariant'
import fs from 'fs-extra'
import path from 'path'

/**
 * Create a path of directories.
 *
 * @param filepath  the give file path
 * @param isDir     Whether the given path is a directory
 */
export function mkdirsIfNotExists(
  filepath: string,
  isDir: boolean,
  logger?: ChalkLogger,
): void {
  const dirPath = isDir ? filepath : path.dirname(filepath)
  if (fs.existsSync(dirPath)) return

  // Print verbose log.
  if (logger != null && logger.verbose != null) {
    logger.verbose(`mkdirs: ${dirPath}`)
  }

  fs.mkdirsSync(dirPath)
}

/**
 * Ensure critical filepath exists, otherwise, kill the process (synchronizing)
 *
 * @param filepath
 */
export function ensureCriticalFilepathExistsSync(
  filepath: string | null,
): void | never {
  let errMsg: string | null = null

  if (filepath == null) {
    errMsg = `Invalid path: ${filepath}.`
  } else if (!fs.existsSync(filepath!)) {
    errMsg = `Not found: ${filepath}.`
  } else if (!fs.statSync(filepath).isFile()) {
    errMsg = `Not a file: ${filepath}.`
  }

  invariant(errMsg == null, errMsg)
}

/**
 * Check whether if the dirPath is a non-existent path or empty folder.
 * (synchronizing)
 *
 * @param dirpath   directory path
 */
export function isNonExistentOrEmpty(dirpath: string | null): boolean {
  if (dirpath == null) return false
  if (!fs.existsSync(dirpath)) return true
  const stat = fs.statSync(dirpath)
  if (!stat.isDirectory()) return false
  const files: string[] = fs.readdirSync(dirpath)
  return files.length <= 0
}

/**
 * Check whether if the dirpath is a directory path. (synchronizing)
 *
 * @param dirpath   directory path
 */
export function isDirectorySync(dirpath: string | null): boolean {
  if (dirpath == null) return false
  if (!fs.existsSync(dirpath)) return false
  const stat = fs.statSync(dirpath)
  return stat.isDirectory()
}

/**
 * Check whether if the filepath is a file path. (synchronizing)
 *
 * @param filepath   file path
 */
export function isFileSync(filepath: string | null): boolean {
  if (filepath == null) return false
  if (!fs.existsSync(filepath)) return false
  const stat = fs.statSync(filepath)
  return stat.isFile()
}
