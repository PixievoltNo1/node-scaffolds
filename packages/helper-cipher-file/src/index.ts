export * from './types/IFileCipher'
export * from './types/IFileCipherBatcher'
export * from './types/IFileCipherFactory'
export * from './util/fingerprint'

export * from './FileCipher'
export * from './FileCipherBatcher'
export * from './FileCipherCatalog'
export * from './FileCipherCatalog.readonly'
export * from './FileCipherFactory'
