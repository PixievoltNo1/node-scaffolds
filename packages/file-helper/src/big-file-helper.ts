import invariant from '@guanghechen/invariant'
import fs from 'fs-extra'
import type { FilePartItem } from './types'
import { consumeStreams } from './util/stream'

export interface BigFileHelperOptions {
  /**
   * The suffix name of a file part.
   * @default '.ghc-part'
   */
  readonly partSuffix?: string

  /**
   * Buffer encoding.
   * @default undefined
   */
  readonly encoding?: BufferEncoding
}

/**
 * Inspired by https://github.com/tomvlk/node-split-file.
 */
export class BigFileHelper {
  public readonly partSuffix: string
  public readonly encoding?: BufferEncoding

  constructor(options: BigFileHelperOptions = {}) {
    this.partSuffix = options.partSuffix ?? '.ghc-part'
    this.encoding = options.encoding
  }

  /**
   * Calculate the name of parts of sourcefile respectively.
   *
   * @param filepath
   * @param parts
   * @returns
   */
  public calcPartFilepaths(filepath: string, parts: FilePartItem[]): string[] {
    if (parts.length <= 1) return [filepath]

    // Part name (file name of part)
    // get the max number of digits to generate for part number
    // ex. if original file is split into 4 files, then it will be 1
    // ex. if original file is split into 14 files, then it will be 2
    // etc.
    const maxPaddingCount = String(parts.length).length

    const partFilepaths = parts.map(part => {
      // construct part number for current file part
      // <file>.sf-part01
      // ...
      // <file>.sf-part14
      const partCode = String(part.sid).padStart(maxPaddingCount, '0')
      return filepath + this.partSuffix + partCode
    })

    return partFilepaths
  }

  /**
   * Split file with part descriptions.
   *
   * @param filepath
   * @param parts
   * @returns
   */
  public async split(filepath: string, parts: FilePartItem[]): Promise<string[]> {
    if (parts.length <= 1) return [filepath]

    const tasks: Array<Promise<void>> = []
    const partFilepaths: string[] = this.calcPartFilepaths(filepath, parts)

    for (let i = 0; i < partFilepaths.length; ++i) {
      const part = parts[i]
      const partFilepath = partFilepaths[i]

      // Create a range in the specified range of the file.
      const reader = fs.createReadStream(filepath, {
        encoding: this.encoding,
        start: part.start,
        end: part.end - 1,
      })

      // Save part
      const task = new Promise<void>((resolve, reject) => {
        const writer = fs.createWriteStream(partFilepath)
        reader.on('error', reject).on('end', resolve).pipe(writer)
      })

      // The operation of splitting the source file can be processed in parallel.
      tasks.push(task)
    }

    await Promise.all(tasks)
    return partFilepaths
  }

  /**
   * Merge files
   *
   * @param inputFilepaths
   * @param outputFilepath
   */
  public async merge(inputFilepaths: string[], outputFilepath: string): Promise<void> {
    invariant(inputFilepaths.length > 0, 'Input file list is empty!')

    const streams: fs.ReadStream[] = inputFilepaths.map(filepath =>
      fs.createReadStream(filepath, {
        encoding: this.encoding,
      }),
    )

    const writer: fs.WriteStream = fs.createWriteStream(outputFilepath, {
      encoding: this.encoding,
    })

    // The operation of merging files could not be processed in parallel.
    await consumeStreams(streams, writer)

    writer.close()
  }
}

export const bigFileHelper = new BigFileHelper()
