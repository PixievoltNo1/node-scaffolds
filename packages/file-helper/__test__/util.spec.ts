import { locateFixtures } from 'jest.setup'
import { calcFilePartItemsByCount, calcFilePartItemsBySize } from '../src'

describe('calcFilePartItemsByCount', function () {
  test('basic', function () {
    const filepath = locateFixtures('a.md')
    expect(calcFilePartItemsByCount(filepath, 4)).toMatchSnapshot()
  })
})

describe('calcFilePartItemsBySize', function () {
  test('basic', function () {
    const filepath = locateFixtures('a.md')
    expect(calcFilePartItemsBySize(filepath, 1024)).toMatchSnapshot()
  })
})
