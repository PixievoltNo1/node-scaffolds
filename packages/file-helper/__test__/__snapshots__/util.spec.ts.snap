// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calcFilePartItemsByCount basic 1`] = `
Array [
  Object {
    "end": 1336,
    "sid": 1,
    "start": 0,
  },
  Object {
    "end": 2672,
    "sid": 2,
    "start": 1336,
  },
  Object {
    "end": 4008,
    "sid": 3,
    "start": 2672,
  },
  Object {
    "end": 5341,
    "sid": 4,
    "start": 4008,
  },
]
`;

exports[`calcFilePartItemsBySize basic 1`] = `
Array [
  Object {
    "end": 1024,
    "sid": 1,
    "start": 0,
  },
  Object {
    "end": 2048,
    "sid": 2,
    "start": 1024,
  },
  Object {
    "end": 3072,
    "sid": 3,
    "start": 2048,
  },
  Object {
    "end": 4096,
    "sid": 4,
    "start": 3072,
  },
  Object {
    "end": 5120,
    "sid": 5,
    "start": 4096,
  },
  Object {
    "end": 5341,
    "sid": 6,
    "start": 5120,
  },
]
`;
